content/blog/2019-12-18-bril-osr/index.md

```diff
@@ -0,0 +1,373 @@
+follows the structure of getting a variable from the stack, applying some
+operation to it, and then storing back on the stack. For example, the assembly
+for an instruction `a: int = add b c` would look roughly like:
+
+```
+mov rax, [rbp - b_stack_position]
+add rax, [rbp - c_stack_position]
+mov [rbp - a_stack_position], rax
+```
+
+In order to implement the `print` instruction, we created basic Rust functions
+to print an integer or a boolean, and then used standard calling conventions to
+call these functions from the assembly code.
+
+[function prologue]: https://en.wikipedia.org/wiki/Function_prologue
+[`dynasm-rs`]: https://github.com/CensoredUsername/dynasm-rs
+
+#### Function Calls
+
+The most difficult part of code generation was implementing function calls (as
+defined in the Bril() language extension). Because this is a just-in-time
+compiler, a function being called may or may not already be compiled. As such,
+we cannot simply call the other function using standard calling conventions.
+Instead, we created another Rust function, `handle_call` for the assembly to
+call when encountering a Bril `call` instruction. To this function, the assembly
+passes an integer identifying the Bril function that it wants to call.
+
+This function can then see if the Bril function has already been compiled, and
+run it if so, or otherwise decide whether to compile it now or interpret it.
+However, to perform these actions, `handle_call` needs access to the internal
+state of the compiler, and, as such, really needs to be a method of the main
+compiler struct. And, to call a method, you need to pass in a reference to the
+`self` variable. As such, the assembly Bril function needs to have a reference
+to this struct. To solve this, we decided on a convention that every time an
+assembly function is called, it is passed a reference to `self` as an argument.
+That reference is then stored on the stack for later use.
+
+Handling function arguments was quite tricky, as Bril() functions can have any
+arbitrary number of arguments, and the Rust `handle_call` function must be able
+to pass along these arguments. Because Rust does not support functions with
+variable numbers of arguments, we decided to pass the Bril arguments in a Rust
+vector to `handle_call`. It took considerable effort to figure out how to create
+a vector from assembly, and pass it as an argument. Unable to find any
+definitive information on how vectors are implemented, we resorted to
+reverse-engineering vector uses. We heavily used the [Godbolt][] compiler
+explorer to quickly see the assembly generated from Rust functions. We made
+assumptions about how vectors work based on the outputs we saw, and based our
+implementation of function arguments on those assumptions. Having done this, we
+were able to construct Rust vectors from assembly and pass them to the
+`handle_call` function, which then forwards them to the callee Bril function to
+be deconstructed.
```

Comment:
That does look like it would have made things much easier! Unfortunately, for most of us this was our first experience with Rust, and so we were unfamiliar with the FFI 😭.